### PR TITLE
feat(createstream): add conditional rendering to CreateStream component

### PR DESCRIPTION
**Pull request by alexisnsns**
Display stream modal only if user has 1/enough native token to pay for transaction fees & 2/ enough tokens to pay for the stream. Also make 'open stream' button clickable only if 'approve' button has been clicked.

Fixes: SBK-246

### CODE REVIEW

File: ui/app/src/App.tsx
Comment (fubuloubu) [inline ui/app/src/App.tsx:8]: Are you having a hard time being able to use the package from editable install?

File: ui/app/src/App.tsx
Comment (fubuloubu) [inline ui/app/src/App.tsx:63]: If this is giving you troubles, can probably remove and add in a separate PR

NOTE: the `StreamStatus` component is supposed to help display the amount of stream time left for something like the deployment list page in chaosnet (as one field in the display of each item)

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:29]: `nativeBalance` might be easier to read

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:54]: Just double checking my understanding of react, this will trigger the `useEffect` on every update of `address`?

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:45]: Can this one be performed in parallel to the above by wrapping in a 2nd `useEffect`?

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:32]: can these be renamed to `feeData`, `feeDataIsError` and `feeDataIsLoading`?

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:63]: You're only using `data` here though?

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:127]: `isApproved` should be dynamically checked vs. on-chain data somehow (e.g. `token.allowance(wallet, stream_manager) > 0`)

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:138]: If we do this in parallel:
```suggestion
          <p>Checking gas and native token balance...</p>
```

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:163]: This way it just falls-through if this sequence finishes first (no confusion about "what was step 2/2?")
```suggestion
          <p>Checking stream token balance...</p>
```

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:221]: Wonder if we want to make this a multi-part sequence:
1. Confirm cart and select payment token (TODO item, just USDC for now... maybe make it a dropdown that just displays USDC)
2. ("Confirm" in CreateStream =>) Select desired stream length based on selected token and stream cost
3. ("Approve" in CreateStream => wait for approval transaction =>) Review stream parameters and execute stream creation (which updates via `props.registerStream`)

File: ui/lib/CreateStream.tsx
Comment (alexisnsns) [reply ui/lib/CreateStream.tsx:29]: OK, updated

File: ui/lib/CreateStream.tsx
Comment (alexisnsns) [reply ui/lib/CreateStream.tsx:54]: Yes this is the dependency array; it runs whenever there is a change in the address. Otherwise you'd be able to load the page then change your metamask account.

File: ui/lib/CreateStream.tsx
Comment (alexisnsns) [reply ui/lib/CreateStream.tsx:63]: Good catch!

Comment (alexisnsns) [reply ui/lib/CreateStream.tsx:63]: Was useful for testing but no need in prod

File: ui/app/src/App.tsx
Comment (alexisnsns) [reply ui/app/src/App.tsx:63]: ok

File: ui/lib/CreateStream.tsx
Comment (alexisnsns) [reply ui/lib/CreateStream.tsx:45]: Very good idea; separation of concerns 101; updated

File: ui/lib/CreateStream.tsx
Comment (alexisnsns) [reply ui/lib/CreateStream.tsx:133]: ok dynamically checked now;

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:293]: Would be nicer to display total approved divided by the amount per second to show how many days you paid for

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:290]: Would be nicer to see total approval amount + token symbol

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:286]: Don't know how useful this field is to show

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:132]: this execution is reverting, lets debug

Comment (fubuloubu) [reply ui/lib/CreateStream.tsx:132]: but otherwise this function is great

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [reply ui/lib/CreateStream.tsx:182]: p.s. this would replace `props.tokenAddress` with a tokenlist to select from

Comment (alexisnsns) [reply ui/lib/CreateStream.tsx:182]: I'll create an issue so that it is addressed in another PR

File: ui/app/TokenList.json
Comment (fubuloubu) [inline ui/app/TokenList.json:12]: use sepolia's chain ID - 58008

File: ui/app/TokenList.json
Comment (fubuloubu) [inline ui/app/TokenList.json:10]:
```suggestion
    },
    {
      "chainId": 58008,
      "address": "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
      "name": "USDCoin",
      "decimals": 6,
      "symbol": "USDC"
    },
```

File: ui/app/TokenList.json
Comment (fubuloubu) [inline ui/app/TokenList.json:5]: This is the token on optimism
```suggestion
      "chainId": 10,
```

File: ui/app/TokenList.json
Comment (fubuloubu) [inline ui/app/TokenList.json:9]:
```suggestion
      "symbol": "USDC.e"
```

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [reply ui/lib/CreateStream.tsx:195]: figured it out, need to use promise api

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:191]: move this function to `props` and this function can be in the demo app

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:209]: wonder if it would make sense to move these display variables to the demo app, inside of `props.registerStream`

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:214]: Might make sense to add all this to a `props.streamCreationError` function handler for the demo app to use upstream

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:220]: same here

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:411]: Move to demo app as well

File: ui/app/TokenList.json
Comment (fubuloubu) [inline ui/app/TokenList.json:12]: whoops this is Arbitrum's chain ID
```suggestion
      "chainId": 42161,
```

**Review by fubuloubu — APPROVED** [review]

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:272]: should first filter by chain ID of network you are on

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:42]: add a comment on why this is used to help the user progress through the sequence in scenarios where there is some error

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:48]: can use promise syntax here instead e.g. `fetchBalance(...).then(...).catch(...)`

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:75]: same here

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:99]: add a comment here

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:188]: what happems where `txHash` is `undefined`?

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:224]: usually have some sort of "choose a token" (that's not a selectible option) which it defaults to while loading the full list in the background e.g. this should be an optional token selection

I do think that the tokenlist could be a prop to this component though

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:230]: move this to a prop and pass through in demo app

also, use `@uniswap/token-lists` types here

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:242]: move this to demo app (or just a build-time input for the demo app)

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:377]: is this the token amount? getting `Days paid: 8.64e-14` in the UI

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:389]: Not sure we even need this with the button display

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:405]: if could move each step outside of the rendering here it would be easier to follow

File: ui/lib/CreateStream.tsx
Comment (alexisnsns) [reply ui/lib/CreateStream.tsx:188]: the logic is done on the previous step; `  useEffect(() => {
    if (isSuccess && data?.hash) {
      setTxHash(data?.hash);
    }
  }, [isSuccess, data?.hash]);
`

subsequent function only loads only with isSuccess; otherwise you get  {isError && <div>You did not confirm the transaction</div>}

File: ui/app/TokenList.json
Comment (fubuloubu) [inline ui/app/TokenList.json:19]: sorry, this was the wrong chain ID
```suggestion
      "chainId": 11155111,
```

File: ui/app/TokenList.json
Comment (fubuloubu) [inline ui/app/TokenList.json:30]: fixed above
```suggestion
```

File: ui/app/src/App.tsx
Comment (fubuloubu) [inline ui/app/src/App.tsx:96]: won't need this with the tokenlist update

File: ui/app/src/App.tsx
Comment (fubuloubu) [inline ui/app/src/App.tsx:101]: or this

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:35]: these should be internal state

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:24]: as noted, you're doing this now with tokenlist prop

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:220]: this function keeps giving me a failing transaction in my wallet. will need to debug

File: ui/lib/CreateStream.tsx
Comment (fubuloubu) [inline ui/lib/CreateStream.tsx:224]: nothing to return really

```suggestion
        sm
```

File: ui/app/src/App.tsx
Comment (fubuloubu) [inline ui/app/src/App.tsx:83]: been using the wrong address this whole time
```suggestion
          streamManagerAddress={"0x3543Faeeddb7bAbCbBB216B3627f9c5E0C39CE41"}
```

File: ui/app/src/config.ts
Comment (fubuloubu) [inline ui/app/src/config.ts:1]: Add a comment to note that you should override this in your downstream implementations as needed

**Review by fubuloubu — APPROVED** [review]

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**
